### PR TITLE
Add More Landmark Functionality

### LINKED ISSUE #42: Landmark show page - visitor
Visitor can see landmark's attributes, picture, and map

### LINKED ISSUE #44: Landmark Show Page Touchups
- Add link of landmark's website
- If there is no info for attribute, display 'Not Available'
- Set image size

### PR DESCRIPTION
**Pull request by lpile**
**Changes proposed in this pull request:**
 - Refactor TomTomService with spec
 - Add Google Service with spec
 - Update Landmark model with attributes, relationships, and spec
 - Add LandmarksShowFacade
 - Add rake task to load landmarks database
 - Add Faker, JQuery, and Bootstrap gems
 
Resolves #42
Resolves #44
 
**The following checks have been completed:**
 - [x] Tested my new feature(s) as well as any feasible edge cases (if possible)
 - [x] Checked `coverage/index.html` - did not add any new code that's not covered by testing (if possible)
 - [x] Merged in the latest master to my branch with `git pull origin master` & resolved merge conflicts
 - [x] Ran `rails db:migrate`
 - [x] Ran the test suite - all tests are passing (or maybe skipped)
 - [x] Poked around in localhost with a small screen size - my new features (as well as basic functionality of the site) are working
 
**Notes:**
 - I will add landmark create_or_find_by on another branch

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -60,6 +60,10 @@ gem "omniauth-rails_csrf_protection"
 # print different colors to the terminal
 gem 'colorize'
 
+# Bootstrap
+gem 'jquery-rails'
+gem 'bootstrap-sass'
```
Comment (chakeresa) [inline Gemfile]: Why do you need this gem? We already had bootstrap installed the conventional way, and things are rendering very weird on a phone on this branch...

File: app/assets/javascripts/landmarks.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+function initMap(lat, lng) {
+    var myCoords = new google.maps.LatLng(lat, lng);
+    var mapOptions = {
+    center: myCoords,
+    zoom: 17
+    };
+    var map = new google.maps.Map(document.getElementById('map'), mapOptions);
```
Comment (chakeresa) [inline app/assets/javascripts/landmarks.js]: The map is not working in my localhost (see my Slack message / screenshot). Let's make sure everything is configured right before we merge this / push to Heroku.

File: app/facades/landmarks_show_facade.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  def initialize(id)
+    @id = id
+  end
+
+  def landmark
+    @_landmark ||= Landmark.find(@id)
+  end
+
+  def picture
+    "https://maps.googleapis.com/maps/api/place/photo?maxheight=400&maxwidth=600&key=#{ENV['GOOGLE_PLACES_API_KEY']}&photoreference=#{landmark.photo_reference}"
```
Comment (chakeresa) [inline app/facades/landmarks_show_facade.rb]: These aren't working for me -- I get a huge default image for several different places

File: app/views/landmarks/show.html.erb [hunk 4]
```diff
@@ -1,16 +1,75 @@
...
+<% provide :head_tags do %>
+  <meta name='turbolinks-visit-control' content='reload'>
+  <script>
+    document.addEventListener("DOMContentLoaded", function(){
+      initMap(<%=facade.landmark.lat%>, <%=facade.landmark.long%>)
+    });
+  </script>
 <% end %>
+
+<!DOCTYPE html>
```
Comment (chakeresa) [inline app/views/landmarks/show.html.erb]: What's going on here? None of the stuff below should be in the view itself. Maybe part of it needs to go in the head in `application.html.erb`.

File: app/views/shared/_nav.html.erb [hunk 5]
```diff
@@ -1,37 +1,33 @@
-<nav class="navbar navbar-expand-sm navbar-dark bg-dark">
-  <%= link_to "Oddio", landmarks_path, class: "navbar-brand" %>
-  <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarsExample03" aria-controls="navbarsExample03" aria-expanded="false" aria-label="Toggle navigation">
+<nav class="navbar navbar-expand-lg navbar-dark bg-dark">
```
Comment (chakeresa) [inline app/views/shared/_nav.html.erb]: The nav bar is very messed up for me on mobile... I'd rather not make these changes. I already setup bootstrap for all that.

File: app/views/shared/_nav.html.erb [hunk 6]
```diff
@@ -1,37 +1,33 @@
...
-          <a class="btn btn-primary my-2 my-sm-0" href="<%= login_path %>">Login</a>
+          <a class="nav-link" href="<%= login_path %>">Login</a>
         </li>
         <li class="nav-item">
-          <a class="btn btn-primary my-2 my-sm-0" href="<%= register_path %>">Register</a>
+          <a class="nav-link" href="<%= register_path %>">Register</a>
         </li>
       <% end %>
     </ul>
+    <form class="form-inline my-2 my-lg-0 pull-right" action="<%= landmarks_path %>">
```
Comment (chakeresa) [inline app/views/shared/_nav.html.erb]: May want to keep this part though -- your search box looked better than the original.

File: lib/tasks/add_landmarks.rake [hunk 7]
```diff
@@ -0,0 +1,78 @@
+require 'digest/md5'
```
Comment (chakeresa) [inline lib/tasks/add_landmarks.rake]: I still need to review this file -- TBD

File: spec/factories/landmarks.rb [hunk 8]
```diff
@@ -1,10 +1,14 @@
...
+    lat { Faker::Address.latitude }
+    long { Faker::Address.longitude }
+    address { Faker::Address.full_address }
+    phone_number { Faker::PhoneNumber.phone_number }
     category { "park" }
-    sequence(:phone_number) { |n| "Phone #{n}" }
+    place_id { Faker::Crypto.sha1 }
+    website { Faker::Internet.url }
+    photo_reference { 'CmRaAAAA58Iy6Em8ucPC_-OCxKTJ-p7YKDBUQ4mgVZj4qlFN6j2JhXKSfteprF-Ckca73TjXaYUtNk-trEbhBCWoUcWWsO42KRap_rlw4teWz4hqBU57sAoTl3-myl_4xjOw3LmBEhBTUAINq_SeOPXlwudc8ccKGhSH-pCQxOO6rRE-O1zFKPW1-h3OqA' }
+    md5_hash { Faker::Crypto.md5 }
```
Comment (chakeresa) [inline spec/factories/landmarks.rb]: In the next branch (when we use the md5_hash), we should make this factory generate the real hash of the data.

File: spec/services/google_service_spec.rb [hunk 9]
```diff
@@ -0,0 +1,63 @@
+require "rails_helper"
```
Comment (chakeresa) [inline spec/services/google_service_spec.rb]: Still need to review this file -- TBD

File: app/services/google_service.rb [hunk 10]
```diff
@@ -0,0 +1,44 @@
+class GoogleService
```
Comment (chakeresa) [inline app/services/google_service.rb]: Still need to review this file -- TBD

File: spec/services/tom_tom_service_spec.rb [hunk 11]
```diff
@@ -1,13 +1,34 @@
 require "rails_helper"
```
Comment (chakeresa) [inline spec/services/tom_tom_service_spec.rb]: Still need to review this file -- TBD

File: Gemfile (hunk 1 above)
Comment (lpile) [reply Gemfile]: Sorry, I did not see it. We can remove them from the Gemfile.

File: app/facades/landmarks_show_facade.rb (hunk 3 above)
Comment (lpile) [reply app/facades/landmarks_show_facade.rb]: Putting the Google API key in the application.yml and also updating the landmark table should fix this issue.

File: app/views/shared/_nav.html.erb (hunk 5 above)
Comment (lpile) [reply app/views/shared/_nav.html.erb]: I'm not attached to the changes I've made to the navbar.

File: app/views/landmarks/show.html.erb (hunk 4 above)
Comment (lpile) [reply app/views/landmarks/show.html.erb]: Sorry, I forgot to get rid of this. I'll clean it up.

File: app/views/shared/_nav.html.erb (hunk 5 above)
Comment (lpile) [reply app/views/shared/_nav.html.erb]: The navbar is better when I removed the html stuff out of the view.

File: spec/factories/landmarks.rb (hunk 8 above)
Comment (lpile) [reply spec/factories/landmarks.rb]: Agree

File: app/views/shared/_nav.html.erb (hunk 5 above)
Comment (lpile) [reply app/views/shared/_nav.html.erb]: Nav bar should be working properly with the removal of the html syntax out of the view.

File: spec/factories/landmarks.rb (hunk 8 above)
Comment (lpile) [reply spec/factories/landmarks.rb]: I've updated the factory to represent the real hash.

**Comment by lpile** [comment]
I've pushed up the corrections to most of the issues. Pull down the branch again, bundle, reseed landmarks table, and make sure the google api key is in and everything should be working. Let me know if you come across anything else.

File: app/assets/javascripts/landmarks.js (hunk 2 above)
Comment (lpile) [reply app/assets/javascripts/landmarks.js]: Try reloading landmarks table.

File: app/facades/landmarks_show_facade.rb (hunk 3 above)
Comment (lpile) [reply app/facades/landmarks_show_facade.rb]: Try reloading landmarks table.

File: app/assets/javascripts/landmarks.js (hunk 2 above)
Comment (chakeresa) [reply app/assets/javascripts/landmarks.js]: Map works after running updated rake task!

File: spec/services/google_service_spec.rb (hunk 9 above)
Comment (chakeresa) [reply spec/services/google_service_spec.rb]: Reviewed

File: spec/services/tom_tom_service_spec.rb (hunk 11 above)
Comment (chakeresa) [reply spec/services/tom_tom_service_spec.rb]: Reviewed

**Comment by chakeresa** [comment]
Resolves #10 
Resolves #9

**Pull request merged**